substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageBFDNameProvider.java

```diff
@@ -375,22 +412,113 @@ public String bfdMangle(Member m) {
+         * demangles to Foo. The second binding demangles to Foo*.
+         *
+         * n.b. repeated use of the pointer operator results in repeated bindings. So, if a
+         * parameter with type Foo** were to be encoded as PP3Foo three bindings would recorded
+         * which demangle to Foo, Foo* and Foo**. However, multiple indirections do not occur in
+         * Java signatures.
+         *
+         * n.b.b. repeated mentions of the same symbol would redundantly record a binding. For
+         * example if int Hello::compareTo(Hello*) were mangled to _ZN5Hello9compareToEJiP5Hello the
+         * resulting bindings would be S_ ==> Hello, S0_ ==> Hello::compareTo, S1_ ==> Hello and S2_
+         * ==> Hello* i.e. both S_ and S1_ would demangle to Hello. This situation should never
+         * arise. A name can always be correctly encoded without repeats. In the above example that
+         * would be _ZN5Hello9compareToEJiPS_.
+         */
         final NativeImageBFDNameProvider nameProvider;
         final StringBuilder sb;
-        final List<String> prefixes;
+
+        // A list of lookup names identifying substituted elements. A prospective name for an
+        // element that is about to be encoded can be looked up in this list. If a match is found
+        // the list index can be used to identify the relevant short symbol. If it is not found
+        // then inserting the name serves to allocate the short name associated with the inserted
+        // element's result index.
+        List<LookupName> bindings;
+
+        /**
+         * A lookup name is used as a key to record and subsequently lookup a short symbol that can
+         * replace one or more elements of a mangled name. A lookup name is usually just a simple
+         * string, i.e. some text that is to be mangled and to which the corresponding short symbol
+         * should decode. However, substitutions for namespace prefixes (e.g. AppLoader506::Foo)
+         * require a composite lookup name that composes a namespace prefix (AppLoader506) with a
+         * trailing simple name (Foo). The corresponding short symbol will demangle to the string
+         * produced by composing the prefix and simple name with a :: separator (i.e. restoring
+         * AppLoader506::Foo). Short symbols for pointer types (e.g. Foo* or AppLoader506::Foo*)
+         * require a pointer lookup name that identifies the substituted text as a pointer to some
+         * underlying type (e.g. Foo or AppLoader506::Foo). The corresponding short symbol will
+         * demangle to the string produced by demangling the underlying type with a * suffix. In
+         * theory the prefix for a pointer type lookup name might be defined by another pointer type
+         * lookup name (if, say, we needed to encode type Foo**). In practice, that case should not
+         * arise with Java method signatures.
+         */
+        private interface LookupName {
```

Comment:
These two interfaces could turned into sealed interfaces.
(Although not needed right now as you do not switch over them)

See https://docs.oracle.com/en/java/javase/17/language/sealed-classes-and-interfaces.html at _Declaring Sealed Interfaces_.

Reply:
I have sealed both interfaces.